fix(wasm): remove type defs for `ArrayExpressionElement` and `Elision`

overlookmotel opened the pull request:
Follow-on after #6404.

`ArrayExpressionElement` and `Elision` are not used in the TS types, because `ArrayExpression` has an override for the field it uses.

https://github.com/oxc-project/oxc/blob/002289b4b1b0d6fac080c4b12f4939c8f455272f/crates/oxc_ast/src/ast/js.rs#L293-L302

Prevent these TS type defs being emitted by introducing a new `#[estree(custom_ts_def)]` attr, to go with `#[estree(custom_serialize)]`.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6683?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#6683** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6683?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#6682** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6682?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6681** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6681?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6680** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6680?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6679** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6679?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6678** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6678?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6676** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6676?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6675** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6675?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6674** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6674?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6673** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6673?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* **#6672** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/6672?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
@ottomated This is the most substantial change in this stack of PRs. The change is correct, but implementation is maybe a bit hacky. You may see a better way to achieve this.

ottomated commented on tasks/ast_tools/src/derives/estree.rs:
```diff
@@ -25,12 +25,17 @@ impl Derive for DeriveESTree {
-        let ts_type_def = quote! {
-            #[wasm_bindgen::prelude::wasm_bindgen(typescript_custom_section)]
-            const TS_APPEND_CONTENT: &'static str = #ts_type_def;
+        let ts_type_def = if let Some(ts_type_def) = ts_type_def {
```
Maybe Option::map_or would make this easier to read? I think the implementation is fine especially since it'll be temporary.

overlookmotel replied:
My personal preference is more explicit code, rather than the sometimes obscure `Option` methods. But that's just a matter of personal taste. Going to leave it as it is for now - I don't think a big deal either way.

overlookmotel committed:
fix(wasm): remove type defs for `ArrayExpressionElement` and `Elision` (#6683)

Follow-on after #6404.

`ArrayExpressionElement` and `Elision` are not used in the TS types, because `ArrayExpression` has an override for the field it uses.

https://github.com/oxc-project/oxc/blob/002289b4b1b0d6fac080c4b12f4939c8f455272f/crates/oxc_ast/src/ast/js.rs#L293-L302

Prevent these TS type defs being emitted by introducing a new `#[estree(custom_ts_def)]` attr, to go with `#[estree(custom_serialize)]`.
